src/pip/_internal/req/__init__.py

```diff
@@ -60,39 +71,95 @@ def install_given_reqs(
             ', '.join(to_install.keys()),
         )
 
-    installed = []
+    # pre allocate installed package names
+    installed = collections.OrderedDict({name: None for name in to_install})
+
+    install_args = [install_options, global_options, dict(
+        root=root, home=home, prefix=prefix,
+        warn_script_location=warn_script_location,
+        use_user_site=use_user_site, pycompile=pycompile)]
 
     with indent_log():
-        for req_name, requirement in to_install.items():
-            if requirement.should_reinstall:
-                logger.info('Attempting uninstall: %s', req_name)
-                with indent_log():
-                    uninstalled_pathset = requirement.uninstall(
-                        auto_confirm=True
-                    )
-            else:
-                uninstalled_pathset = None
-
-            try:
-                requirement.install(
-                    install_options,
-                    global_options,
-                    root=root,
-                    home=home,
-                    prefix=prefix,
-                    warn_script_location=warn_script_location,
-                    use_user_site=use_user_site,
-                    pycompile=pycompile,
-                )
-            except Exception:
-                # if install did not succeed, rollback previous uninstall
-                if uninstalled_pathset and not requirement.install_succeeded:
-                    uninstalled_pathset.rollback()
-                raise
-            else:
-                if uninstalled_pathset and requirement.install_succeeded:
-                    uninstalled_pathset.commit()
-
-            installed.append(InstallationResult(req_name))
-
-    return installed
+        # first try to install in parallel
+        installed_pool = _safe_pool_map(
+            partial(_single_install, install_args, in_subprocess=True),
+            list(to_install.values()))
+        if installed_pool:
+            installed = collections.OrderedDict(
+                zip(list(to_install.keys()), installed_pool))
```

Comment:
This implementation relies on the dict ordering being stable (i.e. subsequent `dict.keys()` and `dict.values()` calls always return entries in the same order), which I believe is true for CPython, but not guaranteed for other Python implementations. I would try to only iterate once instead to be safe.

BTW the comment reads unfinished?

Reply:
Yes that makes sense, I'll refactor it to use the original requirements list, the OrderedDict is actually only used for outputting logging information, and should be avoided if we want to guarantee installation order.